cmd/internal/cli/build_linux.go

```diff
@@ -33,44 +33,69 @@ import (
 	"github.com/apptainer/apptainer/pkg/runtime/engine/config"
 	"github.com/apptainer/apptainer/pkg/sylog"
 	"github.com/apptainer/apptainer/pkg/util/cryptkey"
-	"github.com/apptainer/apptainer/pkg/util/namespaces"
 	keyClient "github.com/apptainer/container-key-client/client"
 	"github.com/spf13/cobra"
 )
 
-func fakerootExec(cmdArgs []string) {
-	if buildArgs.nvccli && !buildArgs.noTest {
-		sylog.Warningf("Due to writable-tmpfs limitations, %%test sections will fail with --nvccli & --fakeroot")
-		sylog.Infof("Use -T / --notest to disable running tests during the build")
-	}
-
-	useSuid := starter.IsSuidInstall()
-
+func fakerootExec() {
+	// First remove fakeroot option from args and environment if present
 	short := "-" + buildFakerootFlag.ShortHand
 	long := "--" + buildFakerootFlag.Name
-	envKey := fmt.Sprintf("APPTAINER_%s", buildFakerootFlag.EnvKeys[0])
-	fakerootEnv := os.Getenv(envKey) != ""
-
-	argsLen := len(os.Args) - 1
-	if fakerootEnv {
-		argsLen = len(os.Args)
-		os.Unsetenv(envKey)
+	for _, pfx := range env.ApptainerPrefixes {
+		envKey := fmt.Sprintf("%s_%s", pfx, buildFakerootFlag.EnvKeys[0])
+		if os.Getenv(envKey) != "" {
+			os.Unsetenv(envKey)
+		}
 	}
-	args := make([]string, argsLen)
-	idx := 0
+	var args []string
 	for i, arg := range os.Args {
 		if i == 0 {
 			path, _ := osExec.LookPath(arg)
 			arg = path
 		}
+		// This does not treat options before the "build" command
+		//   differently than after, which is OK currently because
+		//   there is no -f defined there
 		if arg != short && arg != long {
-			args[idx] = arg
-			idx++
+			if len(arg) > 2 && arg[0] == '-' && arg[1] != '-' {
+				// remove all f within the multiple short options
+				arg = strings.ReplaceAll(arg, buildFakerootFlag.ShortHand, "")
+				if arg == "-" {
+					// could have been -ff
+					continue
+				}
+			}
+			args = append(args, arg)
 		}
+	}
 
+	var err error
+	uid := uint32(os.Getuid())
+	if uid != 0 && !fakeroot.IsUIDMapped(uid) {
+		sylog.Infof("User not listed in %v, trying root-mapped namespace", fakeroot.SubUIDFile)
+		os.Setenv("_APPTAINER_FAKEFAKEROOT", "1")
+		err = fakeroot.UnshareRootMapped(args)
+		if err == nil {
+			// All the work has been done by the child process
+			os.Exit(0)
+		}
+		sylog.Debugf("%v", err)
```

Comment:
Need some debug context

Reply:
And shouldn't it be a call to `sylog.Fatal` at this point ?

Reply:
I thought the message from the child function would be sufficient, but I'll go ahead and add context.

It isn't fatal at this point because it hasn't yet tried the fakeroot command.  That happens later in runBuild(), and then if it hasn't yet become some kind of user id == 0 and fakeroot isn't found, at that point it fails.  The only weird thing about it was the message about --fix-perms comes before that check.  I'll put a comment here and move the message about fix-perms to after when fakeroot is found if uid != 0.

Reply:
Oh, but in testing it just now I realized that if UnshareRootmapped fails that also implies that a suid installation is needed in order to run a build, so I'll add a check for that there too, to fail earlier even before looking for the fakeroot command.